crypto/s390xcap.c

```diff
@@ -178,9 +209,151 @@ void OPENSSL_cpuid_setup(void)
         OPENSSL_s390xcap_P.kdsa[0] &= cap.kdsa[0];
         OPENSSL_s390xcap_P.kdsa[1] &= cap.kdsa[1];
     }
+
+#if defined(OPENSSL_SYS_LINUX) && !defined(FIPS_MODULE)
+    if (cex == 0) {
+        OPENSSL_s390xcex = -1;
+    } else {
+        OPENSSL_s390xcex = open("/dev/z90crypt", O_RDWR | O_CLOEXEC);
+        if (probe_cex() == 1)
+            OPENSSL_atexit(OPENSSL_s390x_cleanup);
+    }
+#endif
+}
+
+#if defined(OPENSSL_SYS_LINUX) && !defined(FIPS_MODULE)
+static int probe_cex(void)
+{
+    struct ica_rsa_modexpo me;
+    const unsigned char inval[16] = {
+                                     0,0,0,0,0,0,0,0,
+                                     0,0,0,0,0,0,0,2
+    }, modulus[16] = {
+                      0,0,0,0,0,0,0,0,
+                      0,0,0,0,0,0,0,3
+    };
+    unsigned char res[16];
+    int olderrno;
+    int rc = 1;
+
+    me.inputdata = (unsigned char *)inval;
+    me.inputdatalength = sizeof(inval);
+    me.outputdata = (unsigned char *)res;
+    me.outputdatalength = sizeof(res);
+    me.b_key = (unsigned char *)inval;
+    me.n_modulus = (unsigned char *)modulus;
+    olderrno = errno;
+    if (ioctl(OPENSSL_s390xcex, ICARSAMODEXPO, &me) == -1) {
+        (void)close(OPENSSL_s390xcex);
+        OPENSSL_s390xcex = -1;
+        rc = 0;
+    }
+    errno = olderrno;
+    return rc;
 }
 
-static int parse_env(struct OPENSSL_s390xcap_st *cap)
+static int s390x_mod_exp_hw(BIGNUM *r, const BIGNUM *a, const BIGNUM *p,
```

Comment:
Please do not put the s390x_mod_exp_hw() and related functions into s390xcap.c. Instead it should be in crypto/bn subdirectory in a new file and added to build.info similarly to bn_ppc.c.

Reply:
Done.